### PR TITLE
Re-enable HMD Comfort Mode

### PR DESCRIPTION
**Pull request by hyperlogic**
There is now a HMD Comfort Mode Enabled setting, available in the "Settings > Avatar..." menu.  This is enabled by default.  Disabling this setting will allow the avatar to be pushed when attempting to lean over tables or walk through walls.

### CODE REVIEW

**Comment by hyperlogic** [comment]
QA test plan with HMD Comfort Mode Enabled:

* In HMD, when standing on the ground, movement via the drive keys, (using the keyboard or vive touchpad) has linear velocity, and instantaneously moves the avatar as the drive keys are pressed.
* Ensure that flying functions.
* Ensure that flying works with collisions disabled.
* In room scale VR if you walk (not drive) into a table or a wall, your camera will enter the geometry, but your avatar will be pushed back by the obstruction.  This will separate your camera view from your avatar's head.

QA Test plan with HMD Comfort Mode Disabled:

* In HMD when you drive using the keyboard or hand controllers joystick movement no longer has linear velocity, and instead will quickly accelerate and decelerate as the drive keys are pressed.
* Ensure that flying functions.
* Ensure that flying works with collisions disabled.
* In room scale VR if you walk (not drive) into a table or a wall, your avatar will be pushed back by the obstruction.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14e7da8f9900dd4e01264557a2e1c6569683387e available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-14e7da8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-14e7da8.dmg

File: interface/src/ui/PreferencesDialog.cpp [hunk 1]
```diff
@@ -80,6 +80,11 @@ void setupPreferences() {
         preference->setStep(0.25);
         preferences->addPreference(preference);
     }
+    {
+        auto getter = [=]()->bool { return myAvatar->getHMDComfortModeEnabled(); };
+        auto setter = [=](bool value) { myAvatar->setHMDComfortModeEnabled(value); };
```
Comment (sethalves) [inline interface/src/ui/PreferencesDialog.cpp]: should this one be & ?

Comment (hyperlogic) [inline interface/src/ui/PreferencesDialog.cpp]: I admit I just copied the pattern from earlier code in this method.  But, It is safer,  because it copies a shared_ptr, which will insure that myAvatar will never be destroyed until the lambda is.  I don't think this will lead to a memory leak, because Preferences and MyAvatar are singletons anyway.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e11a33f4e60077cdef36340e2a44ead670f9c5c7 available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-e11a33f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-e11a33f.dmg

**Comment by hyperlogic** [comment]
I found a bug with non HMD Comfort Mode, which causes the world rotate slightly as you move rapidly in room scale.  I'm holding off QA testing until I have a fix for this.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f90ce6c3ba95bc6746973ba4e0e791773927870f available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-f90ce6c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-f90ce6c.dmg

File: libraries/physics/src/CharacterController.cpp [hunk 2]
```diff
@@ -480,42 +483,58 @@ void CharacterController::setParentVelocity(const glm::vec3& velocity) {
...
+                // use simple blending to filter noise from the velocity measurement
+                const float blend = 0.2f;
+                _followVelocity = (1.0f - blend) * _followVelocity + blend * newFollowVelocity;
+            }
+        }
+        _previousFollowPosition = _followDesiredBodyTransform.getOrigin();
+        _followTimeAccumulator = 0.0f;
     }
-    _previousFollowPosition = _followDesiredBodyTransform.getOrigin();
-    _followTimeAccumulator = 0.0f;
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterController.cpp]: Not a blocker: Clearing _`followTimeAccumulator` needs to be done for both the `if` and `else` cases so it should stay down here.

File: libraries/physics/src/CharacterController.cpp [hunk 3]
```diff
@@ -480,42 +483,58 @@ void CharacterController::setParentVelocity(const glm::vec3& velocity) {
...
-            _followVelocity = newFollowVelocity;
-            newFollowVelocity.setY(0.0f);
-            _rigidBody->setLinearVelocity(newFollowVelocity);
+    if (!_hmdComfortModeEnabled) {
+
+        // directly teleport to the desired position & orientation, but keep track of the velocity for animation purposes.
+
+        glm::vec3 followTranslation = extractTranslation(desiredWorldBodyMatrix);
+        if (_followTimeAccumulator > 0.0f) {
+            _followVelocity = (glmToBullet(followTranslation) - _previousFollowPosition) / _followTimeAccumulator;
```
Comment (AndrewMeadows) [inline libraries/physics/src/CharacterController.cpp]: I know this wouldn't be a new problem with this PR but I thought I would mention it since it occurred to me...

I wonder, since the math here is done in the world-frame... if there happened to be a "teleport" mid follow would that cause a very large velocity to be computed here?  That is, if `_previousFollowPosition` were from before the teleport but `followTranslation` were from after?

I tried to audit the code to determine if this theoretical problem is real or not but could not manage it while using the git diff format.  Probably not a blocker (the bug would be rare) but maybe we should investigate more thoroughly someday and look for a place where we need to update ` _previousFollowPosition` on teleport events.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 12e7dd84232e00c3e0b92b22a978a25cd32981bf available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-12e7dd8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-12e7dd8.dmg

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by Nex-Pro** [comment]
Tested on win10 and Vive two users.

### FAIL

We both experienced difficulty flying.

in HMD mode with comfort on the avatar flicks into and out of view ahead of the path of motion. You see either the inside of the head or even the back then the avatar falls behind the view. This is a continuous repeated action. There is a sensation that you are fighting to correct the heading(yaw). This makes flying less than enjoyable.

In desktop mode the drift issue raises it's head again.  I could adjust my yaw to be more or less zero then press forward on the keyboard. Without touching any other controls the heading would begin to yaw until eventually the heading (yaw)  was effectively 180. This was repeatable but not 100% consistent.
The yaw would come on gradually building to a large number of degrees per second then fall off as 180 was approached.  I am not sure what this indicates code wise.

In comfort mode Michelle was able to trigger a rapid vertical acceleration in my avatar up to a height of more than 15 meters by swiping me rapidly side to side with a croquet mallet.
I was able to return the compliment in a reasonably repeatable manner.


### QA test plan with HMD Comfort Mode Enabled:

In HMD, when standing on the ground, movement via the drive keys, (using the keyboard or vive touchpad) has linear velocity, and instantaneously moves the avatar as the drive keys are pressed. **SUCCESS**

Ensure that flying functions, and stopping is instantaneous.  **More or less.**

Ensure that flying works with collisions disabled. **VERY POOR, FAIL**

In room scale VR if you walk (not drive) into a table or a wall, your camera will enter the geometry, but your avatar will be pushed back by the obstruction. This will separate your camera view from your avatar's head. **SUCCESS**

### QA Test plan with HMD Comfort Mode Disabled:

In HMD when you drive using the keyboard or hand controllers joystick movement no longer has linear velocity, and instead will quickly accelerate and decelerate as the drive keys are pressed. **SUCCESS**

Ensure that flying functions. **VERY POOR, FAIL**

Ensure that flying works with collisions disabled. **VERY POOR, FAIL**

In room scale VR if you walk (not drive) into a table or a wall, your avatar will be pushed back by the obstruction. **SUCCESS**

**Additional:**  Enable disable comfort mode in avatar settings dialog :  **SUCCESS**

**Comment by Nex-Pro** [comment]
@hyperlogic  2x Thanks for adding the option to disable comfort mode.

**Comment by OmegaHeron** [comment]
Windows 10

When in Desktop mode with this feature enabled AV is prone to being uncontrollable while flying using keyboard or xbox stick controls.  Giving left/right turning input AV will corkscrew and become very difficult to control.  Turn comfort mode off and all is well again.

Same with DK2 HMD.  Well, same but worse in that going into the deathspin with HMD on is far less pleasant than desktop...  3rd person view of AV flying with DK2 on exhibits behavior Nex-pro described - it jerks looking like it's going to fly apart.

Bottom line.  I would not want this on while in Desktop mode or using DK2.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d7a7d8774890dd24be62bd13cd3c99c5e84e6490 available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-d7a7d87.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-d7a7d87.dmg

**Comment by hyperlogic** [comment]
@Nex-Pro, I need more detailed information about the following items:

1. In HMD with comfort enabled, Ensure that flying works with collisions disabled.
2. In HMD with comfort disabled, Ensure that flying functions.
3. In HMD with comfort disabled, Ensure that flying works with collisions disabled.

You tagged each of these as VERY POOR FAIL. Without any description of what issues you found.  From my perspective as a developer, 1 works as intended and 2 & 3 work almost identically to how flying works in master.  Can you provide more information please.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4696bbbc868c14f33e1023353c043f7dc2a0fbf5 available at: https://gist.github.com/64b730db0318fa5512b085fa7d1b6e03<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-4696bbb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9166-4696bbb.dmg

**Comment by Nex-Pro** [comment]
### FAIL

In HMD, when standing on the ground, movement via the drive keys, (using the keyboard or vive touchpad) has linear velocity, and instantaneously moves the avatar as the drive keys are pressed. **SUCCESS**

Ensure that flying functions, and stopping is instantaneous. **FAIL** avatar can pass through horizontal collidable surfaces in the down direction. ( negative y )

Ensure that flying works with collisions disabled. **SUCCESS**

In room scale VR if you walk (not drive) into a table or a wall, your camera will enter the geometry, but your avatar will be pushed back by the obstruction. This will separate your camera view from your avatar's head.


QA Test plan with HMD Comfort Mode Disabled:

In HMD when you drive using the keyboard or hand controllers joystick movement no longer has linear velocity, and instead will quickly accelerate and decelerate as the drive keys are pressed.  **SUCCESS**

Ensure that flying functions.

Ensure that flying works with collisions disabled. **SUCCESS**

In room scale VR if you walk (not drive) into a table or a wall, your avatar will be pushed back by the obstruction. **SUCCESS**

In addition with comfort mode enabled the avatar flicks in and out of view in the HMD.
